File: library/src/main/java/renetik/java/collections/CSMap.java
package renetik.java.collections;

import renetik.android.lang.CSLang;

public interface CSMap<K, V> extends java.util.Map<K, V> {
    static <T, D> D value(CSMap<T, D> map, T key) {
        if (CSLang.no(map)) return null;
        return map.value(key);
    }

    @Deprecated
    boolean containsKey(Object key);

    @Deprecated
    boolean containsValue(Object value);

    V get(Object key);

    boolean hasKey(K key);

    boolean hasValue(V value);

    V value(K key);

}

File: library/renetik-base/src/main/java/renetik/android/java/collections/CSMap.kt
package renetik.android.java.collections

fun <K, V> map() = CSMapImpl<K, V>()
fun <K, V> linkedMap() = CSLinkedMapImpl<K, V>()

fun map(vararg values: Any) = map<Any, Any>().apply {
    var i = 0
    while (i < values.size) {
        put(values[i], values[i + 1])
        i += 2
    }
}

fun map(vararg values: String) = map<String, String>().apply {
    var i = 0
    while (i < values.size) {
        put(values[i], values[i + 1])
        i += 2
    }
}

interface CSMap<K, V> : Map<K, V>, MutableMap<K, V> {

    fun hasKey(key: K): Boolean

    fun hasValue(value: V): Boolean

    fun value(key: K): V?

}